### ISSUE TITLE
Unhandled NPE during reconciliation

### ISSUE THREAD

**Issue by serrss** [strimzi/strimzi#239]
## **Step to reproduce:**

1. Clone and open Strimzi sources https://github.com/strimzi/strimzi
2. Up cluster
`$ oc clucter up`
3. Login as admin
`$ oc login -u system:admin`
4. Deploy Cluster Controller
`$ oc create -f resources/openshift/cluster-controller-with-template.yaml`
5. Deploy Topic Controller
`$ oc create -f resources/openshift/topic-controller-with-template.yaml`
6. Login to OpenShift https://127.0.0.1:8443
7. Create **"Apache Kafka (Persistent storage)"**
7.1 Click "Add to project"
7.2 Click "Select from project"
7.3 Select "Apache Kafka (Persistent storage)" and click "Next"
7.4 Use default configurations:
_Number of Zookeper cluster nodes: 3
Number of Kafka cluster nodes: 3_
7.5 Click "Create" and wait for the message "Apache Kafka (Persistent storage) has been created"
7.6 Open page "Pods" and wait when all pods will have status "Running"
8. Create **"Strimzi Topic Controller"**
8.1 Click "Add to project"
8.2 Click "Select from project"
8.3 Select "Strimzi Topic Controller" and click "Next"
8.4 Use default configurations
8.5 Click "Create" and wait for the message "Success Strimzi Topic Controller has been created."
8.6 Open page "Pods" and wait when all pods will have status "Running"
9. Open page "Config Maps" and open your cluster config
10. Edit numbers of kafka-nodes
10.1 Click "Actions"
10.2 Click "Edit YAML"
10.3 Set new value (1) for _variable kafka-nodes_ and save
11. Open page "Pods" and open logs for _strimzi-topic-controller_ and wait 2-3 mins

## **Actual result:**
Log has an unhandled exception:
```
[2018-02-22 14:01:58,135] WARN  <etworkClient:241> [dminclient-1] [AdminClient clientId=adminclient-1] Connection to node 2 could not be established. Broker may not be available.
--
  | [2018-02-22 14:02:01,206] WARN  <etworkClient:241> [dminclient-1] [AdminClient clientId=adminclient-1] Connection to node 1 could not be established. Broker may not be available.
  | [2018-02-22 14:07:47,274] INFO  <Session     :158> [oop-thread-0] Starting periodic reconciliation
  | Feb 22, 2018 2:07:47 PM io.vertx.core.impl.ContextImpl
  | SEVERE: Unhandled exception
  | java.lang.NullPointerException
  | at io.strimzi.controller.topic.Controller.lambda$reconcile$1(Controller.java:342)
  | at io.vertx.core.impl.CompositeFutureImpl.lambda$all$1(CompositeFutureImpl.java:54)
  | at io.vertx.core.impl.FutureImpl.tryFail(FutureImpl.java:170)
  | at io.vertx.core.impl.FutureImpl.fail(FutureImpl.java:102)
  | at io.vertx.core.impl.FutureImpl.handle(FutureImpl.java:154)
  | at io.vertx.core.impl.FutureImpl.handle(FutureImpl.java:23)
  | at io.strimzi.controller.topic.ZkTopicStore.lambda$read$1(ZkTopicStore.java:71)
  | at io.strimzi.controller.topic.zk.ZkImpl.lambda$null$4(ZkImpl.java:199)
  | at io.vertx.core.impl.ContextImpl.lambda$wrapTask$2(ContextImpl.java:344)
  | at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
  | at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:403)
  | at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:463)
  | at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
  | at java.lang.Thread.run(Thread.java:748)
```

**Comment by ppatierno** [comment]
@serrss I tried but what I see after 15 minutes (x 2) is the following : 

```
[2018-03-08 15:34:18,474] WARN  <etworkClient:241> [dminclient-1] [AdminClient clientId=adminclient-1] Connection to node 1 could not be established. Broker may not be available.
[2018-03-08 15:43:01,744] INFO  <Session     :158> [oop-thread-0] Starting periodic reconciliation
[2018-03-08 15:58:01,740] INFO  <Session     :158> [oop-thread-0] Starting periodic reconciliation
```

so no exception :-(

It's clear thet in the `Controller` at line 342 the `ar.result()` is null as reason for the NPE you saw (so something failed) but I cannot reproduce it right now.

**Comment by ppatierno** [comment]
I was able to reproduce the issue but creating a topic before scaling down Kafka is mandatory otherwise no reconcile is started and the code where the mentioned NPE should happen is not involved. I am going into it.

**Comment by ppatierno** [comment]
So the problem is the following ...

In the steps provided by @serrss it's needed to create a new topic using a ConfigMap for example (so let's say 8.1 step). Only after that it's possible to scale down the Kafka broker for having the issue.

Scaling down the broker drives the Admin Client having some problems on reconnecting to the last alive broker (maybe it's another issue). If you delete the topic ConfigMap after scaling down the topic, the reconciliation will start.

Tha Admin Client starts to try sending metadata request to the Kafka cluster but this request times out.
So in the `Controller` around the line 342, the `topicStore.read` went well but the `kafka.topicMetadata` went wrong with the following exception : 

Timed out waiting to send the call (TimeoutException)

It means that the `ar.failed()` is true and `ar.result()` is null.

To fix the NPE we should check that `ar.failed()` is false before doing what it's done today and eventually tracing the Admin Client exception as well.

In any case the fact that the Admin Client stops to work properly when the cluster is scaled down should be treated as a separate issue.

**Comment by ppatierno** [comment]
Closed by #305

**Comment by ppatierno** [comment]
@serrss It should be fixed by #305 . Give it a new try.

**Comment by serrss** [comment]
@ppatierno issue was fixed
